torch/CMakeLists.txt

```diff
@@ -264,6 +268,8 @@ target_link_libraries(torch_python ${TORCH_PYTHON_LINK_LIBRARIES})
 
 target_compile_definitions(torch_python PRIVATE ${TORCH_PYTHON_COMPILE_DEFINITIONS})
 
+target_compile_definitions(torch_python PUBLIC ${TORCH_PYTHON_PUBLIC_COMPILE_DEFINITIONS})
```

Comment:
Not so happy about this, TBH.

The issue is that we guard some parts of the headers with USE_TENSORPIPE, which means that every downstream user that includes those headers must also set the right value of USE_TENSORPIPE. (Note that some "downstream users" are part of PyTorch itself, like the tests). Instead of asking each user to take care to set their USE_TENSORPIPE compile definition to the same value as their USE_TENSORPIPE CMake variable, we could do that for them by adding it to the interface of the CMake target.

We used to do this in TensorPipe itself, in order to allow users to detect what transports/channels had been enabled during compilation, but it came back to bite us, and it's the reason we didn't ship SHM and CMA in PyTorch 1.6. What we settled with now is to define these flags in an auto-generated header file.

However, I'm not aware of any such header file in PyTorch. And, moreover, this approach is what c10d itself is doing to let its users know which of Gloo/NCCL/MPI is enabled:
https://github.com/pytorch/pytorch/blob/fb8aa0046c554630cb9f173df8b4c7c751415312/torch/lib/c10d/CMakeLists.txt#L108-L118
So perhaps it's OK?

Reply:
You can ask cmake/buck to bake the dependency for you in one of generated header, for example by adding `#cmakedefine USE_TENSORPIPE` to https://github.com/pytorch/pytorch/blob/686705c98b43f65e48e3a920b24634e65f54aca1/c10/macros/cmake_macros.h.in#L7-L10

(Although, this header is not the perfect place for that,  but the same is true for other `.h.in` files)

Reply:
Yes, I do agree that having a generated header is more robust, and I guess I'll do that. However, even though c10, ATen and Caffe2 have one, PyTorch doesn't have any generated header yet, so adding one might be a bit of work. I'd like to land this soon-ish as there's some other work blocked by it, so if it's acceptable I'd prefer to go with this for now and introduce the generated header later...